### DISCUSSION TITLE
logging in as admin without a username crashes the server

### DISCUSSION THREAD

**Discussion by archcloudy** [computernewb/collabvm-1.2.ts#11] [General]
Hi, yesterday i wrote my own collabvm bot in node.js.
Today, i added the function to make the bot log in as admin at every startup
but when i ran the bot, collabvm server crashed with this error message:

```
/path/to/collabvm-1.2.ts/build/guacutils.js:28
          command += current.toString().length + '.' + current;

TypeError: Cannot read properties from undefined (reading 'toString')
```

How i did was to send a websocket message, which is `5.admin,1,2,18.password`.
So what is going on, does that websocket message only works in client? (or cvm-webapp)

**Comment by archcloudy** [comment]
sorry i made a typo in the websocket message

`2.admin,1,2,18.password`

**Comment by elijahr2411** [comment]
I'm unable to reproduce this crash with either of the messages you sent. Have you modified CollabVM server at all? Are you running the latest version?

Additionally, both of those messages are invalid syntax. The correct syntax is described [here](https://guacamole.apache.org/doc/0.9.5/gug/guacamole-protocol.html).

**Comment by archcloudy** [comment]
i'm running on latest collabvm server.
and i used this code to send message to the server so that the bot can log in as admin
```
const adminpass = '';
const modpass = '';

function client() {
  console.log(adminpass) // Just to identify whether password is confliting or not. Only temporary
  ws.on('connect', function(wsclient) {
    console.log('Bot connected!');
    
    const message = encodeCommand(['admin', '2', adminpass]);
    console.log(message) // Apperantly logs the encoded websocket message
    wsclient.sendUTF(message)
}

client();
```
![image](https://github.com/computernewb/collabvm-1.2.ts/assets/86261006/b91a814b-0b46-4aca-978e-01aa75e86752)

**Comment by archcloudy** [comment]
i rechecked with the guacamole documentation that you've attached there, and i see now that the syntax is incorrect.

So i went ahead and fixed it with this new one:
```
const loginadmin = "5.admin,1.2,4." + adminpass
wsclient.sendUTF(loginadmin,)
```
and it's still doesn't work.

![IMG_20240207_191414](https://github.com/computernewb/collabvm-1.2.ts/assets/86261006/c98b6519-4b4e-4345-ab61-4bed874992f5)
![IMG_20240207_191331](https://github.com/computernewb/collabvm-1.2.ts/assets/86261006/c7f089ab-9605-432e-b485-d83bcc915455)

**Comment by elijahr2411** [comment]
I'm still not able to reproduce the server crash.

Would you mind sending the full source to your bot so I can investigate further?

**Comment by elijahr2411** [comment]
Disregard that, we've found the issue.

The problem is that you're trying to log in as an admin before assigning a username, which crashes the VM

(this is a low-priority bug as it only works if you know the correct admin password)

The crash is fixed by 73ab61f, as any login attempts will now be ignored. To properly log in to the VM, you need to first send a 7.rename instruction.

**Comment by archcloudy** [comment]
sorry i'm kinda confused rn with the "7.rename" instruction thing
![image](https://github.com/computernewb/collabvm-1.2.ts/assets/86261006/c8fb6a0a-0570-4ec0-a9a5-205505c56200)

**Comment by archcloudy** [comment]
sorry just to make it clear, the bot did joined with assigned username

**Comment by archcloudy** [comment]
![image](https://github.com/computernewb/collabvm-1.2.ts/assets/86261006/d85931e0-41bf-4fc7-b935-954851464c62)
i just realized i have that code already

**Comment by coolgamer17223** [comment]
we've tried to use 7.rename but the bot doesnt show up we used 6.rename instead and the bot showed up now we just arent able to login as admin

**Comment by archcloudy** [comment]
we already have that in our code, but the problem is we can't still log in to admin

**Comment by coolgamer17223** [comment]
We've already added the encoded functions. Also our bot's cant login to Admin.

**Comment by archcloudy** [comment]
do you guys need my full cvm code so you guys can get into futher more investigation?

**Comment by elijahr2411** [comment]
At this point the issue with the server has been solved, please create a discussion or a thread in our discord server to discuss problems with your code.

**Comment by ghost** [comment]


**Reply by archcloudy** [reply]
alright thank you

**Comment by archcloudy** [comment]
okay so what's next?

> elijahr2411: At this point the issue with the server has been solved, please create a discussion or a thread in our discord server to discuss problems with your code.

**Reply by coolgamer17223** [reply]
This is my bot's code:
require('dotenv').config();
const WebSocketClient = require('websocket').client;
const OpenAI = require('openai');

const openai = new OpenAI.OpenAIApi({
    apiKey: process.env.OPENAI_API_KEY,
});

class Config {
    constructor() {
        this.conxTimeout = 600;
        this.ip = "ready-system.gl.at.ply.gg:65500";
        this.name = 'TestBot-JS';
        this.prefix = '!';
        this.adminPassword = 'hunter2'; // Consider using an environment variable here
    }
}

var config = new Config();

async function getAIResponse(question) {
    try {
        const response = await openai.createCompletion({
            model: "text-davinci-003", // Adjust model as necessary
            prompt: question,
            max_tokens: 60
        });
        return response.data.choices[0].text.trim();
    } catch (error) {
        console.error("Error in OpenAI response:", error);
        return "Sorry, I couldn't process your question.";
    }
}

function connectToServer() {
    var ws = new WebSocketClient();
    console.log('Connecting to VM1...');

    ws.on('connectFailed', function (error) {
        console.log('Failed to connect: ' + error.toString());
        connectToServer();
    });

    ws.on('connect', function (conx) {
        console.log('Connected to VM1 as ' + config.name + '!');

        function chat(message) {
            conx.sendUTF(encodeCommand(['chat', message]));
        }

        function loginAdmin() {
            conx.sendUTF(encodeCommand(['admin',2, config.adminPassword]));
        }

        // Rename the bot immediately after connection
        function renameBot() {
            const renameCommand = `6.rename,${config.name.length}.${config.name};`;
            conx.sendUTF(renameCommand);
        }

        renameBot();
        loginAdmin();

        conx.on('error', function (error) {
            console.log('Error while connecting: ' + error.toString());
            setTimeout(function () {
                connectToServer();
            }, config.conxTimeout);
        });

        conx.on('close', function () {
            console.log('Connection was closed.');
            setTimeout(function () {
                connectToServer();
            }, config.conxTimeout);
        });

        conx.on('message', async function (message) {
            var cmd = decodeCommand(message.utf8Data);
            var command = cmd[2];
            var prefix = config.prefix;

            if (cmd[0] == 'chat') {
                if (command.startsWith(prefix + "ask ")) {
                    var question = command.slice((prefix + "ask ").length);
                    var answer = await getAIResponse(question);
                    chat(answer);
                } else if (command.startsWith(prefix + "vote")) {
                    chat("You really thought, huh? There's no such vote command.");
                }
                // ... Other command handlers
            }
        });

        setInterval(function () {
            if (conx.connected) {
                conx.sendUTF('3.nop;');
            }
        }, 2500);

        if (conx.connected) {
            chat('Bot is now active.');
        }
    });

    ws.connect('ws://' + config.ip, 'guacamole');
}

function decodeCommand(cypher) {
    var sections = [];
    var bump = 0;

    while (sections.length <= 50 && cypher.length >= bump) {
        var current = cypher.substring(bump);
        var length = parseInt(current.substring(current.search(/\./) - 2), 10);
        var parameter = current.substring(length.toString().length + 1, Math.floor(length / 10) + 2 + length);
        sections[sections.length] = parameter;
        bump += Math.floor(length / 10) + 3 + length;
    }

    sections[sections.length - 1] = sections[sections.length - 1].substring(0, sections[sections.length - 1].length - 1);
    return sections;
}

function encodeCommand(cypher) {
    var command = "";
    for (var i = 0; i < cypher.length; i++) {
        var current = cypher[i];
        command += current.length + "." + current;
        command += (i < cypher.length - 1 ? "," : ";");
    }
    return command;
}

connectToServer();

**Reply by archcloudy** [reply]
@MDMCK10 (just incase you missed it - sorry if that annoyed you)

**Reply by archcloudy** [reply]
Awesome! I'll let you know if the code works.

**Comment by coolgamer17223** [comment]
The bot works thank you!

**Reply by coolgamer17223** [reply]
It also Logins into Admin

**Comment by archcloudy** [comment]
Thank you so much for helping us!
